src/Stratis.Bitcoin.Features.BlockStore/AddressIndexing/AddressIndexer.cs

```diff
@@ -84,205 +109,340 @@ public void Initialize()
+                            await Task.Delay(DelayTimeMs, this.cancellation.Token).ConfigureAwait(false);
+                        }
+                        catch (OperationCanceledException)
+                        {
+                        }
+
+                        continue;
+                    }
+
+                    if (nextHeader.Previous.HashBlock != this.IndexerTip.HashBlock)
+                    {
+                        ChainedHeader lastCommonHeader = nextHeader.FindFork(this.IndexerTip);
+
+                        this.logger.LogDebug("Reorg detected. Rewinding till '{0}'.", lastCommonHeader);
+
+                        lock (this.lockObject)
+                        {
+                            foreach (List<AddressBalanceChange> changes in this.addressesIndex.AddressChanges.Values)
+                                changes.RemoveAll(x => x.BalanceChangedHeight > lastCommonHeader.Height);
+                        }
+
+                        this.IndexerTip = lastCommonHeader;
+                        continue;
+                    }
+
+                    // Get next header block and process it.
+                    Block blockToProcess = this.blockStore.GetBlock(nextHeader.HashBlock);
+
+                    if (blockToProcess == null)
+                    {
+                        this.logger.LogDebug("Next block wasn't found. Waiting.");
+
+                        try
+                        {
+                            await Task.Delay(DelayTimeMs, this.cancellation.Token).ConfigureAwait(false);
+                        }
+                        catch (OperationCanceledException)
+                        {
+                        }
+
+                        continue;
+                    }
+
+                    bool success = this.ProcessBlock(blockToProcess, nextHeader);
+
+                    if (!success)
+                    {
+                        this.logger.LogDebug("Failed to process next block. Waiting.");
+
+                        try
+                        {
+                            await Task.Delay(DelayTimeMs, this.cancellation.Token).ConfigureAwait(false);
+                        }
+                        catch (OperationCanceledException)
+                        {
+                        }
+
+                        continue;
+                    }
+
+                    this.IndexerTip = nextHeader;
+
+                    lock (this.lockObject)
+                    {
+                        this.addressesIndex.TipHashBytes = this.IndexerTip.HashBlock.ToBytes();
+                    }
+                }
+
+                lock (this.lockObject)
+                {
+                    this.dataStore.Update(this.addressesIndex);
+                }
             }
-            catch (Exception ex)
+            catch (Exception e)
             {
-                this.logger.LogError(ex.ToString());
-                throw;
+                this.logger.LogCritical(e.ToString());
```

Comment:
When the AsyncProvider is impleented and we ask a task from it, this can throw (@MithrilMan FYI) we should probably add the ability to get a backround task frmo the provider for this scenarios.

Reply:
@dangershony we have IAsyncLoop for that already in AsyncProvider

Reply:
AsyncLoop is limited, so we need to return a Task and that can throw (then redirect to the fault handler)

Reply:
I'm pretty sure it's not limited for our needs, IAsyncLoop exposes its Task, so you can create continuation on it to catch exceptions if needed, let's have a talk about that